File: src/MotionAnalyser.cpp
#include "MotionAnalyser.hpp"
#include <vector>
#include <tuple>
#include <deque>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

/* Definition: CMotionAnalyser */

int CMotionAnalyser::generate_blocks(tuple<int, int> size, int block_size)
{
    for (int y{}; y < std::get<1>(size) - block_size; y += block_size)
    {
        for (int x{}; x < std::get<0>(size) - block_size; x += block_size)
        {
            int w{block_size}, h{block_size};
            if (x + block_size > get<0>(size))
            {
                w = get<0>(size) - x;
            }
            if (y+block_size>get<1>(size))
            {
                h = get<1>(size) - y;
            }
            _blocks.push_back(make_tuple(x, y, w, h));
        }
    }
    return 0;
}

int CMotionAnalyser::initialize_detect_object(const int frame_sample_count,
                                              const double Dist2Threshold,
                                              const bool detectShadows)
{
    /* [ATTENTION] please add some control.*/
    _bgsubtrator = createBackgroundSubtractorKNN(_frame_sample_count, 800, false);
}

int CMotionAnalyser::detect_motion(Mat &output)
{
    if (!_bgsubtrator)
        return 0;

    Mat thres, dila;
    threshold(_bgMask, thres, 244, 255, THRESH_BINARY);
    erode(thres, thres, getStructuringElement(MORPH_ELLIPSE, Size(3, 3)), Point(-1, -1), 2);
    dilate(thres, dila, getStructuringElement(MORPH_ELLIPSE, Size(3, 3)), Point(-1, -1), 2);

    dila.copyTo(output);
    return 1;
}

int CMotionAnalyser::get_motion_blocks(vector<tuple<int, int, int, int>> &blockList,
                                       const Mat &motion_map,
                                       vector<tuple<int, int, int, int>> &destBlockList)
{
    Mat int_diff;
    float t11{}, t12{}, t21{}, t22{};
    int x{}, y{}, w{}, h{};
    integral(motion_map, int_diff);
    for (auto it{blockList.begin()}; it != blockList.end(); it++)
    {
        tie(x, y, w, h) = *it;
        t11 = int_diff.at<float>(y, x);
        t12 = int_diff.at<float>(y, x + w);
        t21 = int_diff.at<float>(y + h, x);
        t22 = int_diff.at<float>(y + w, x + h);
        if (t22 - t12 - t21 + t11 > 0)
        {
            destBlockList.push_back(*it);
        }
    }
    return 1;
}

int CMotionAnalyser::feed_img(const Mat &img, const bool auto_apply)
{
    if (!_bgsubtrator)
    {
        return 0;
    }

    _frameList.push_back(img);
    if (_frameList.size() > _frame_sample_count)
    {
        _frameList.pop_front();
    }

    if (auto_apply)
        _bgsubtrator->apply(img, _bgMask);
    return 1;
}

int CMotionAnalyser::apply_img()
{
    if (!_bgsubtrator)
    {
        return 0;
    }

    for (auto it{_frameList.begin()}; it != _frameList.end(); it++)
    {
        _bgsubtrator->apply(*it, _bgMask);
    }
    return 1;
}

/* Definition: CMotionAnalyserOpticalFlow */

vector<Point2f> CMotionAnalyserOpticalFlow::detect_motion_optical_flow(const cv::Mat &img_mat,
                                                                       const cv::Mat &pre_img_mat,
                                                                       const int block_size)
{
    Mat mgrey, mpre_grey;
    /* Convert image to grey */
    cvtColor(img_mat, mgrey, CV_BGR2GRAY);
    cvtColor(pre_img_mat, mpre_grey, CV_BGR2GRAY);

    vector<Point2f> curt_points, prev_points;
    vector<unsigned char> status;
    vector<float> err;
    vector<Point2f> moving_points;

    goodFeaturesToTrack(mgrey, prev_points, 500, 0.001, 10);
    calcOpticalFlowPyrLK(mpre_grey, mgrey, prev_points, curt_points, status, err, Size(10, 10), 3, TermCriteria(TermCriteria::MAX_ITER + TermCriteria::EPS, 20, 0.03));

    {
        auto curt_it{curt_points.begin()};
        auto prev_it{prev_points.begin()};
        int c{};
        while (curt_it != curt_points.end())
        {
            if (status[c++] && (2 < (abs(prev_it->x - curt_it->x) + abs(prev_it->y - curt_it->y))))
            {
                moving_points.push_back(*curt_it);
            }
            curt_it++;
            prev_it++;
        }
    }

    return moving_points;
}

File: motion_analyser.py
import cv2 as cv

class MotionAnalyser:
    _blocks = []
    _bg_subtractor = None
    _bg_mask = None
    _frame_list = []
    _frame_sample_num = 16

    def __init__(self):
        pass

    # generate blocks' range
    # image_size: (tuple) size of image
    # block_size: (tuple) size of each block
    # return None
    def generate_blocks(self, image_size, block_size):
        image_width, image_height = image_size
        for y in range(0, image_height - block_size, block_size):
            for x in range(0, image_width - block_size, block_size):
                w = image_width - x if block_size + x > image_width else block_size
                h = image_height - y if block_size + y > image_height else block_size
                self._blocks.append((x, y, w, h))

    # initialize KNN Subtractor
    # return None
    def initialize_detect_object(self,
                                 frame_sample_num,
                                 dist_to_threshold,
                                 detect_shadows):
        self._frame_sample_num = frame_sample_num
        self._bg_subtractor = cv.createBackgroundSubtractorKNN(frame_sample_num,
                                                               dist_to_threshold,
                                                               detect_shadows)
        # self._bg_subtractor = cv.bgsegm.createBackgroundSubtractorMOG(frame_sample_num)
        return

    # return motion mask
    def detect_motion(self):
        if self._bg_subtractor is None:
            return 0
        ret, thresh = cv.threshold(self._bg_mask, 244, 255, cv.THRESH_BINARY)
        thresh = cv.erode(thresh, cv.getStructuringElement(cv.MORPH_ELLIPSE, (3, 3), (-1, -1)), iterations=2)
        thresh = cv.dilate(thresh, cv.getStructuringElement(cv.MORPH_ELLIPSE, (3, 3), (-1, -1)), iterations=2)
        return thresh

    # return motion block list
    def get_motion_blcoks(self, motion_map):
        intgral_image = cv.integral(motion_map)
        result_list = []
        for block in self._blocks:
            x, y, w, h = block
            t11 = intgral_image[y, x]
            t12 = intgral_image[y, x + w]
            t21 = intgral_image[y + h, x]
            t22 = intgral_image[y + w, x + h]
            if t22 - t12 - t21 + t11 > 0:
                result_list.append(block)
        return result_list

    # depend on is_auto apply return a copy of bg_mask or None
    def feed_image(self, image, is_auto_apply):
        if self._bg_subtractor is None:
            return 0
        self._frame_list.append(image)
        if len(self._frame_list) > self._frame_sample_num:
            self._frame_list.pop(0)
        if is_auto_apply:
            self._bg_mask = self._bg_subtractor.apply(image)
            return self._bg_mask.copy()
        return

    def apply_img(self):
        if self._bg_subtractor is None:
            return 0
        for frame in self._frame_list:
            self._bg_mask = self._bg_subtractor.apply(frame)
        return self._bg_mask.copy()
